File: src/wrapper/cpp/int_array.cpp
///////////////////////////////////////
//             DSL_INT_ARRAY          //
///////////////////////////////////////

void * createIntArray()
{
    DSL_intArray * arr = new DSL_intArray();
    void * retval = arr;
    return retval;
}
void * createIntArray_InitialSize(int initialSize)
{
    DSL_intArray * arr = new DSL_intArray(initialSize);
    void * retval = arr;
    return retval;
}
void * createIntArray_Copy(void * void_intarr)
{
    DSL_intArray * likeThisOne = reinterpret_cast<DSL_intArray*>(void_intarr);
    DSL_intArray * arr = new DSL_intArray(*likeThisOne);
    void * retval = arr;
    return retval;
}

void freeIntArray(void * void_intarr)
{
    DSL_intArray * arr = reinterpret_cast<DSL_intArray*>(void_intarr);
    delete arr;
}

void intarray_UseAsList(void * void_intarr, int nItems)
{
    DSL_intArray * arr = reinterpret_cast<DSL_intArray*>(void_intarr);
    arr->UseAsList(nItems);
}
int intarray_NumItems(void * void_intarr)
{
    DSL_intArray * arr = reinterpret_cast<DSL_intArray*>(void_intarr);
    return arr->NumItems();
}
int intarray_GetSize(void * void_intarr)
{
    DSL_intArray * arr = reinterpret_cast<DSL_intArray*>(void_intarr);
    return arr->GetSize();
}
int intarray_GetIndex(void * void_intarr, int index)
{
    DSL_intArray * arr = reinterpret_cast<DSL_intArray*>(void_intarr);
    return (*arr)[index];
}
void intarray_SetIndex(void * void_intarr, int index, int value)
{
    DSL_intArray * arr = reinterpret_cast<DSL_intArray*>(void_intarr);
    (*arr)[index] = value;
}
int intarray_Insert(void * void_intarr, int here, int thisNumber)
{
    DSL_intArray * arr = reinterpret_cast<DSL_intArray*>(void_intarr);
    return arr->Insert(here, thisNumber);
}
int intarray_Add(void * void_intarr, int thisNumber)
{
    DSL_intArray * arr = reinterpret_cast<DSL_intArray*>(void_intarr);
    return arr->Add(thisNumber);
}

File: src/low_level/intarray.jl
export
	   to_native_int_array,
       num_items,
       use_as_list

function to_native_int_array(arr::IntArray)
    n = num_items(arr)
    retval = Array(Int32, n)
    for i = 1 : n
        retval[i] = arr[i-1]
    end
    retval
end

use_as_list(arr::IntArray, nItems::Integer=-1) =
    ccall( (:intarray_UseAsList, LIB_SMILE), Void, (Ptr{Void},Cint), arr.ptr, nItems)
num_items(arr::IntArray) = 
    ccall( (:intarray_NumItems, LIB_SMILE), Int32, (Ptr{Void},), arr.ptr )
Base.size(arr::IntArray) = 
    ccall( (:intarray_GetSize, LIB_SMILE), Int32, (Ptr{Void},), arr.ptr )
Base.getindex(arr::IntArray, index::Integer) = 
    ccall( (:intarray_GetIndex, LIB_SMILE), Int32, (Ptr{Void},Int32), arr.ptr, index )
Base.setindex!(arr::IntArray, value::Integer, index::Integer) = 
    ccall( (:intarray_SetIndex, LIB_SMILE), Void, (Ptr{Void},Int32,Int32), arr.ptr, index, value )
Base.insert!(arr::IntArray, here::Integer, thisNumber::Integer) = 
    ccall( (:intarray_Insert, LIB_SMILE), Cint, (Ptr{Void},Cint,Cint), arr.ptr, here, thisNumber )
Base.push!(arr::IntArray, thisNumber::Integer) =
    ccall( (:intarray_Add, LIB_SMILE), Cint, (Ptr{Void},Cint), arr.ptr, thisNumber )